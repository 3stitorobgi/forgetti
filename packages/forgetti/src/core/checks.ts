import * as t from '@babel/types';
import type * as babel from '@babel/core';
import type { ComponentNode, StateContext } from './types';

export function getImportSpecifierName(specifier: t.ImportSpecifier): string {
  if (specifier.imported.type === 'Identifier') {
    return specifier.imported.name;
  }
  return specifier.imported.value;
}

export function isComponent(node: t.Node): node is ComponentNode {
  switch (node.type) {
    case 'ArrowFunctionExpression':
    case 'FunctionExpression':
    case 'FunctionDeclaration':
      return true;
    default:
      return false;
  }
}

export function isComponentNameValid(
  ctx: StateContext,
  node: ComponentNode,
  checkName: boolean,
): boolean {
  if (checkName) {
    if (node.type !== 'ArrowFunctionExpression') {
      return (
        !!node.id
        && (
          ctx.filters.component.test(node.id.name)
          || (!!ctx.filters.hook && ctx.filters.hook.test(node.id.name))
        )
      );
    }
    return false;
  }
  return true;
}

type TypeFilter<V extends t.Node> = (node: t.Node) => node is V;

export function isPathValid<V extends t.Node>(
  path: unknown,
  key: TypeFilter<V>,
): path is babel.NodePath<V> {
  return key((path as babel.NodePath).node);
}

export type NestedExpression =
  | t.ParenthesizedExpression
  | t.TypeCastExpression
  | t.TSAsExpression
  | t.TSSatisfiesExpression
  | t.TSNonNullExpression
  | t.TSInstantiationExpression
  | t.TSTypeAssertion;

export function isNestedExpression(node: t.Node): node is NestedExpression {
  switch (node.type) {
    case 'ParenthesizedExpression':
    case 'TypeCastExpression':
    case 'TSAsExpression':
    case 'TSSatisfiesExpression':
    case 'TSNonNullExpression':
    case 'TSTypeAssertion':
    case 'TSInstantiationExpression':
      return true;
    default:
      return false;
  }
}

/** Check whether a Node is CallExpression and a React Hook call */
export function isExpressionAndHook(
  ctx: StateContext,
  node: t.Expression | t.SpreadElement,
): boolean {
  switch (node.type) {
    case 'CallExpression':
      // A simple check to see if "node.callee" is an Identifier
      // TypeScript can infer the type of "node.callee" to Identifier or V8IntrinsicIdentifier
      if (!('name' in node.callee)) {
        return false;
      }

      // Check if callee is a react hook
      if (!ctx.filters.hook) {
        return false;
      }
      return ctx.filters.hook.test(node.callee.name);
    case 'SpreadElement':
      return isExpressionAndHook(ctx, node.argument);
    case 'ArrayExpression':
      return node.elements.some((element) => {
        if (!element) return false;
        return isExpressionAndHook(ctx, element);
      });
    case 'ObjectExpression':
      return node.properties.some((property) => {
        if (!property) return false;
        if (property.type === 'ObjectProperty') {
          return (
            (t.isExpression(property.value) && isExpressionAndHook(ctx, property.value))
            || (t.isSpreadElement(property.key) && isExpressionAndHook(ctx, property.key))
          );
        }
        if (property.type === 'SpreadElement') {
          return isExpressionAndHook(ctx, property);
        }
        return false;
      });
    case 'TemplateLiteral':
      return node.expressions.some((expression) => {
        if (!expression) return false;
        if (!t.isExpression(expression)) return false;
        return isExpressionAndHook(ctx, expression);
      });
    case 'BinaryExpression':
      return (
        (t.isExpression(node.left) && isExpressionAndHook(ctx, node.left))
      || (t.isExpression(node.right) && isExpressionAndHook(ctx, node.right))
      );
    case 'AssignmentExpression':
      return (t.isExpression(node.right) && isExpressionAndHook(ctx, node.right));
    default:
      return false;
  }
}
