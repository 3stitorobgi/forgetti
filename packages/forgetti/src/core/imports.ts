import type { ImportRegistration } from './presets';

export const RUNTIME_EQUALS: ImportRegistration = {
  name: '$$equals',
  source: 'forgetti/runtime',
  kind: 'named',
};

export const RUNTIME_CACHE: ImportRegistration = {
  name: '$$cache',
  source: 'forgetti/runtime',
  kind: 'named',
};
