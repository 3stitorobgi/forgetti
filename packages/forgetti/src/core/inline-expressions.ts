/* eslint-disable @typescript-eslint/no-unnecessary-condition */
import type * as babel from '@babel/core';
import type { ComponentNode } from './types';

function isInValidExpression(path: babel.NodePath): boolean {
  let current = path.parentPath;
  let prev = path;
  while (current) {
    if (
      current.isConditionalExpression()
      && (
        current.get('consequent').node === prev.node
        || current.get('alternate').node === prev.node
      )
    ) {
      return true;
    }
    if (
      current.isLogicalExpression()
      && current.get('right').node === prev.node
    ) {
      return true;
    }
    prev = current;
    current = current.parentPath;
  }
  return false;
}

export function inlineExpressions(
  path: babel.NodePath<ComponentNode>,
): void {
  path.traverse({
    Expression(p) {
      if (p.getFunctionParent() === path && p.isIdentifier()) {
        const binding = p.scope.getBinding(p.node.name);
        if (binding && binding.referenced && binding.referencePaths.length === 1) {
          switch (binding.kind) {
            case 'const':
            case 'let':
            case 'var': {
              // move the node to the reference
              const ref = binding.referencePaths[0];
              if (
                binding.path.isVariableDeclarator()
                && binding.path.get('id').isIdentifier()
                && binding.path.scope.getBlockParent() === ref.scope.getBlockParent()
                && binding.path.node.init
                && !isInValidExpression(ref)
              ) {
                ref.replaceWith(binding.path.node.init);
                binding.path.remove();
              }
            }
              break;
            default:
              break;
          }
        }
      }
    },
  });
  path.scope.crawl();
}
