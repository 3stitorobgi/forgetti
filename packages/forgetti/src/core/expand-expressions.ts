/* eslint-disable @typescript-eslint/no-unnecessary-condition */
import * as t from '@babel/types';
import type * as babel from '@babel/core';
import type { ComponentNode, StateContext } from './types';
import { isPathValid } from './checks';
import unwrapNode from './unwrap-node';

export function expandExpressions(
  ctx: StateContext,
  path: babel.NodePath<ComponentNode>,
): void {
  if (path.node.type === 'ArrowFunctionExpression') {
    if (path.node.body.type !== 'BlockStatement') {
      path.node.body = t.blockStatement(
        [t.returnStatement(path.node.body)],
      );
    }
  }

  const hoistedVars = new Set();

  path.traverse({
    AssignmentExpression(p) {
      const parent = p.getFunctionParent();

      if (parent === path) {
        if (
          p.parentPath.isVariableDeclarator()
          && p.parentPath.node.id.type === 'Identifier'
          && hoistedVars.has(p.parentPath.node.id.name)
        ) {
          return;
        }
        const statement = p.getStatementParent();
        if (statement) {
          const id = p.scope.generateUidIdentifier('hoisted');
          hoistedVars.add(id.name);
          statement.insertBefore(
            t.variableDeclaration(
              'let',
              [t.variableDeclarator(id, p.node)],
            ),
          );
          p.replaceWith(id);
        }
      }
    },
    CallExpression(p) {
      const parent = p.getFunctionParent();

      if (parent === path) {
        if (
          p.parentPath.isVariableDeclarator()
          && p.parentPath.node.id.type === 'Identifier'
          && hoistedVars.has(p.parentPath.node.id.name)
        ) {
          return;
        }
        const calleePath = p.get('callee');
        if (isPathValid(calleePath, t.isExpression)) {
          let isHook = false;
          const trueID = unwrapNode(calleePath.node, t.isIdentifier);
          if (trueID) {
            const binding = p.scope.getBindingIdentifier(trueID.name);
            if (binding) {
              const registration = ctx.registrations.hooks.get(binding);
              if (registration) {
                isHook = true;
              }
            }
            if (ctx.filters.hook?.test(trueID.name)) {
              isHook = true;
            }
            // Check if callee is potentially a namespace import
          }
          const trueMember = unwrapNode(calleePath.node, t.isMemberExpression);
          if (
            trueMember
            && !trueMember.computed
            && t.isIdentifier(trueMember.property)
          ) {
            const obj = unwrapNode(trueMember.object, t.isIdentifier);
            if (obj) {
              const binding = path.scope.getBindingIdentifier(obj.name);
              if (binding) {
                const registrations = ctx.registrations.hooksNamespaces.get(binding);
                if (registrations) {
                  let registration: typeof registrations[0];
                  for (let i = 0, len = registrations.length; i < len; i += 1) {
                    registration = registrations[i];
                    if (registration.name === trueMember.property.name) {
                      isHook = true;
                    }
                  }
                }
              }
            }
            if (ctx.filters.hook?.test(trueMember.property.name)) {
              isHook = true;
            }
          }

          if (isHook) {
            const statement = p.getStatementParent();
            if (statement) {
              const id = p.scope.generateUidIdentifier('hoisted');
              hoistedVars.add(id.name);
              statement.insertBefore(
                t.variableDeclaration(
                  'let',
                  [t.variableDeclarator(id, p.node)],
                ),
              );
              p.replaceWith(id);
            }
          }
        }
      }
    },
  });

  path.scope.crawl();
}
