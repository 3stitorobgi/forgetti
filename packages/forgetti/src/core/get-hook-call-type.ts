/* eslint-disable @typescript-eslint/no-unnecessary-condition */
import type * as babel from '@babel/core';
import * as t from '@babel/types';
import { isPathValid } from './checks';
import type { HookIdentity } from './presets';
import unwrapNode from './unwrap-node';
import type { StateContext } from './types';

type HookCallType =
  | HookIdentity
  | 'custom'
  | 'none';

export function getHookCallType(
  ctx: StateContext,
  path: babel.NodePath<t.CallExpression | t.OptionalCallExpression>,
): HookCallType {
  const callee = path.get('callee');
  if (isPathValid(callee, t.isV8IntrinsicIdentifier)) {
    return 'none';
  }
  const trueID = unwrapNode(callee.node, t.isIdentifier);
  if (trueID) {
    const binding = path.scope.getBindingIdentifier(trueID.name);
    if (binding) {
      const registration = ctx.registrations.hooks.get(binding);
      if (registration) {
        return registration.type;
      }
    }
    return ctx.filters.hook && ctx.filters.hook.test(trueID.name) ? 'custom' : 'none';
  }
  // Check if callee is potentially a namespace import
  const trueMember = unwrapNode(callee.node, t.isMemberExpression);
  if (
    trueMember
      && !trueMember.computed
      && t.isIdentifier(trueMember.property)
  ) {
    const obj = unwrapNode(trueMember.object, t.isIdentifier);
    if (obj) {
      const binding = path.scope.getBindingIdentifier(obj.name);
      if (binding) {
        const registrations = ctx.registrations.hooksNamespaces.get(binding);
        if (registrations) {
          let registration: typeof registrations[0];
          for (let i = 0, len = registrations.length; i < len; i += 1) {
            registration = registrations[i];
            if (registration.name === trueMember.property.name) {
              return registration.type;
            }
          }
        }
      }
    }
    return ctx.filters.hook && ctx.filters.hook.test(trueMember.property.name)
      ? 'custom'
      : 'none';
  }
  return 'none';
}
