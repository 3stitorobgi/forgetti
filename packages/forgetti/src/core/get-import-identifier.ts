import * as babel from '@babel/core';
import { addDefault, addNamed } from '@babel/helper-module-imports';
import { ImportRegistration } from './presets';
import { StateContext } from './types';

export default function getImportIdentifier(
  ctx: StateContext,
  path: babel.NodePath,
  definition: ImportRegistration,
) {
  const target = `${definition.source}[${definition.name}]`;
  const current = ctx.hooks.get(target);
  if (current) {
    return current;
  }
  const newID = (definition.kind === 'named')
    ? addNamed(path, definition.name, definition.source)
    : addDefault(path, definition.source);
  ctx.hooks.set(target, newID);
  return newID;
}
