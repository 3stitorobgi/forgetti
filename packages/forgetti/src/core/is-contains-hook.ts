/* eslint-disable @typescript-eslint/no-unnecessary-condition */
/* eslint-disable @typescript-eslint/no-use-before-define */
import type * as babel from '@babel/core';
import * as t from '@babel/types';
import { isNestedExpression, isPathValid } from './checks';
import type OptimizerScope from './optimizer-scope';
import type { ComponentNode, StateContext } from './types';
import unwrapNode from './unwrap-node';

interface OptimizerInstance {
  ctx: StateContext;
  scope: OptimizerScope;
  path: babel.NodePath<ComponentNode>;
}

function isTemplateLiteralContainsHook(
  instance: OptimizerInstance,
  path: babel.NodePath<t.TemplateLiteral>,
): boolean {
  const expressions = path.get('expressions');
  for (let i = 0, len = expressions.length; i < len; i++) {
    const expr = expressions[i];
    if (isPathValid(expr, t.isExpression) && isContainsHook(instance, expr)) {
      return true;
    }
  }
  return false;
}

function isConditionalExpressionContainsHook(
  instance: OptimizerInstance,
  path: babel.NodePath<t.ConditionalExpression>,
): boolean {
  return isContainsHook(instance, path.get('test'));
}

function isBinaryExpressionContainsHook(
  instance: OptimizerInstance,
  path: babel.NodePath<t.BinaryExpression>,
): boolean {
  const left = path.get('left');
  if (isPathValid(left, t.isExpression) && isContainsHook(instance, left)) {
    return true;
  }
  const right = path.get('right');
  if (isPathValid(right, t.isExpression) && isContainsHook(instance, right)) {
    return true;
  }
  return false;
}

function isLogicalExpressionContainsHook(
  instance: OptimizerInstance,
  path: babel.NodePath<t.LogicalExpression>,
): boolean {
  return isContainsHook(instance, path.get('left')) || isContainsHook(instance, path.get('right'));
}

function isUnaryExpressionContainsHook(
  instance: OptimizerInstance,
  path: babel.NodePath<t.UnaryExpression>,
): boolean {
  return isContainsHook(instance, path.get('argument'));
}

function isCallExpressionContainsHook(
  instance: OptimizerInstance,
  path: babel.NodePath<t.CallExpression | t.OptionalCallExpression>,
): boolean {
  const callee = path.get('callee');

  if (isPathValid(callee, t.isExpression)) {
    const trueID = unwrapNode(callee.node, t.isIdentifier);
    if (trueID) {
      const binding = path.scope.getBindingIdentifier(trueID.name);
      if (binding) {
        const registration = instance.ctx.registrations.hooks.get(binding);
        if (registration) {
          return false;
        }
        if (instance.ctx.filters.hook?.test(trueID.name)) {
          return true;
        }
      }
    }
    const trueMember = unwrapNode(callee.node, t.isMemberExpression);
    if (
      trueMember
      && !trueMember.computed
      && t.isIdentifier(trueMember.property)
    ) {
      const obj = unwrapNode(trueMember.object, t.isIdentifier);
      if (obj) {
        const binding = path.scope.getBindingIdentifier(obj.name);
        if (binding) {
          const registrations = instance.ctx.registrations.hooksNamespaces.get(binding);
          if (registrations) {
            for (let i = 0, len = registrations.length; i < len; i += 1) {
              const registration = registrations[i];
              if (registration.name === trueMember.property.name) {
                return false;
              }
            }
          }
        }
      }
      if (instance.ctx.filters.hook?.test(trueMember.property.name)) {
        return true;
      }
    }
    if (isContainsHook(instance, callee)) {
      return true;
    }
  }
  const args = path.get('arguments');
  for (let i = 0, len = args.length; i < len; i++) {
    const arg = args[i];
    if (isPathValid(arg, t.isExpression) && isContainsHook(instance, arg)) {
      return true;
    }
    if (isPathValid(arg, t.isSpreadElement) && isContainsHook(instance, arg.get('argument'))) {
      return true;
    }
  }
  return false;
}

function isArrayExpressionContainsHook(
  instance: OptimizerInstance,
  path: babel.NodePath<t.ArrayExpression | t.TupleExpression>,
): boolean {
  const elements = path.get('elements');
  for (let i = 0, len = elements.length; i < len; i++) {
    const element = elements[i];
    if (isPathValid(element, t.isSpreadElement) && isContainsHook(instance, element.get('argument'))) {
      return true;
    }
    if (isPathValid(element, t.isExpression) && isContainsHook(instance, element)) {
      return true;
    }
  }
  return false;
}

function isObjectExpressionContainsHook(
  instance: OptimizerInstance,
  path: babel.NodePath<t.ObjectExpression | t.RecordExpression>,
): boolean {
  const properties = path.get('properties');
  for (let i = 0, len = properties.length; i < len; i++) {
    const property = properties[i];
    if (isPathValid(property, t.isSpreadElement) && isContainsHook(instance, property.get('argument'))) {
      return true;
    }
    if (isPathValid(property, t.isObjectProperty)) {
      if (property.node.computed) {
        const key = property.get('key');
        if (isPathValid(key, t.isExpression) && isContainsHook(instance, key)) {
          return true;
        }
      }
      const value = property.get('value');
      if (isPathValid(value, t.isExpression) && isContainsHook(instance, value)) {
        return true;
      }
    }
  }
  return false;
}

function isSequenceExpressionContainsHook(
  instance: OptimizerInstance,
  path: babel.NodePath<t.SequenceExpression>,
): boolean {
  const exprs = path.get('expressions');
  for (let i = 0, len = exprs.length; i < len; i++) {
    if (isContainsHook(instance, exprs[i])) {
      return true;
    }
  }
  return false;
}

function isJSXChildrenContainsHook(
  instance: OptimizerInstance,
  path: babel.NodePath<t.JSXFragment | t.JSXElement>,
): boolean {
  const children = path.get('children');
  for (let i = 0, len = children.length; i < len; i++) {
    const child = children[i];
    if (isPathValid(child, t.isJSXElement) && isJSXElementContainsHook(instance, child)) {
      return true;
    }
    if (isPathValid(child, t.isJSXFragment) && isJSXChildrenContainsHook(instance, child)) {
      return true;
    }
    if (isPathValid(child, t.isJSXSpreadChild) && isContainsHook(instance, child.get('expression'))) {
      return true;
    }
    if (isPathValid(child, t.isJSXExpressionContainer)) {
      const expr = child.get('expression');
      if (isPathValid(expr, t.isExpression) && isContainsHook(instance, expr)) {
        return true;
      }
    }
  }
  return false;
}

function isJSXElementContainsHook(
  instance: OptimizerInstance,
  path: babel.NodePath<t.JSXElement>,
): boolean {
  return isJSXChildrenContainsHook(instance, path);
}

function isAssignmentExpressionContainsHook(
  instance: OptimizerInstance,
  path: babel.NodePath<t.AssignmentExpression>,
): boolean {
  const left = path.get('left');
  if (isPathValid(left, t.isExpression) && isContainsHook(instance, left)) {
    return true;
  }
  const right = path.get('right');
  if (isPathValid(right, t.isExpression) && isContainsHook(instance, right)) {
    return true;
  }
  return false;
}

function isTaggedTemplateExpressionContainsHook(
  instance: OptimizerInstance,
  path: babel.NodePath<t.TaggedTemplateExpression>,
): boolean {
  return isContainsHook(instance, path.get('tag'))
    || isContainsHook(instance, path.get('quasi'));
}

export default function isContainsHook(
  instance: OptimizerInstance,
  path: babel.NodePath<t.Expression>,
): boolean {
  if (isPathValid(path, isNestedExpression)) {
    return isContainsHook(instance, path.get('expression'));
  }
  if (isPathValid(path, t.isLiteral)) {
    if (isPathValid(path, t.isTemplateLiteral)) {
      return isTemplateLiteralContainsHook(instance, path);
    }
    return false;
  }
  if (isPathValid(path, t.isIdentifier)) {
    return false;
  }
  if (isPathValid(path, t.isMemberExpression) || isPathValid(path, t.isOptionalMemberExpression)) {
    return false;
  }
  if (isPathValid(path, t.isConditionalExpression)) {
    return isConditionalExpressionContainsHook(instance, path);
  }
  if (isPathValid(path, t.isBinaryExpression)) {
    return isBinaryExpressionContainsHook(instance, path);
  }
  if (isPathValid(path, t.isLogicalExpression)) {
    return isLogicalExpressionContainsHook(instance, path);
  }
  if (isPathValid(path, t.isUnaryExpression)) {
    return isUnaryExpressionContainsHook(instance, path);
  }
  if (
    isPathValid(path, t.isCallExpression)
    || isPathValid(path, t.isOptionalCallExpression)
  ) {
    return isCallExpressionContainsHook(instance, path);
  }
  if (
    isPathValid(path, t.isFunctionExpression)
    || isPathValid(path, t.isArrowFunctionExpression)
  ) {
    return false;
  }
  if (isPathValid(path, t.isAssignmentExpression)) {
    return isAssignmentExpressionContainsHook(instance, path);
  }
  if (isPathValid(path, t.isArrayExpression) || isPathValid(path, t.isTupleExpression)) {
    return isArrayExpressionContainsHook(instance, path);
  }
  if (isPathValid(path, t.isObjectExpression) || isPathValid(path, t.isRecordExpression)) {
    return isObjectExpressionContainsHook(instance, path);
  }
  if (isPathValid(path, t.isNewExpression)) {
    return false;
  }
  if (isPathValid(path, t.isSequenceExpression)) {
    return isSequenceExpressionContainsHook(instance, path);
  }
  if (isPathValid(path, t.isTaggedTemplateExpression)) {
    return isTaggedTemplateExpressionContainsHook(instance, path);
  }
  if (isPathValid(path, t.isJSXFragment)) {
    return isJSXChildrenContainsHook(instance, path);
  }
  if (isPathValid(path, t.isJSXElement)) {
    return isJSXElementContainsHook(instance, path);
  }
  if (isPathValid(path, t.isImport)) {
    return false;
  }
  if (isPathValid(path, t.isMetaProperty)) {
    return false;
  }
  return false;
}
