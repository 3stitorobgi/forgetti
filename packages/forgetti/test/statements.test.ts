/* eslint-disable no-template-curly-in-string */
import * as babel from '@babel/core';
import { describe, expect, it } from 'vitest';
import plugin, { Options } from '../src';

const options: Options = {
  componentFilter: {
    source: '^[A-Z]',
    flags: '',
  },
  hookFilter: {
    source: '^use[A-Z]',
    flags: '',
  },
  memo: {
    name: 'useMemo',
    source: 'react',
    kind: 'named',
  },
  hooks: [
    {
      type: 'memo',
      name: 'useMemo',
      source: 'react',
      kind: 'named',
    },
    {
      type: 'callback',
      name: 'useCallback',
      source: 'react',
      kind: 'named',
    },
    {
      type: 'effect',
      name: 'useEffect',
      source: 'react',
      kind: 'named',
    },
    {
      type: 'effect',
      name: 'useLayoutEffect',
      source: 'react',
      kind: 'named',
    },
  ],
  hocs: [],
};

async function compile(code: string) {
  const result = await babel.transformAsync(code, {
    plugins: [
      [plugin, options],
    ],
    parserOpts: {
      plugins: [
        'jsx',
      ],
    },
  });

  return result?.code ?? '';
}

describe('statements', () => {
  it('should optimize for-of statements', async () => {
    const code = `
  function Example(props) {
    for (const x of props.arr) {
      console.log(x);
    }
  }
  `;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize for-in statements', async () => {
    const code = `
  function Example(props) {
    for (const x in props.arr) {
      console.log(x);
    }
  }
  `;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize for statements', async () => {
    const code = `
  function Example(props) {
    for (let i = 0; i < 10; i += 1) {
      console.log(i);
    }
  }
  `;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize while statements', async () => {
    const code = `
  function Example(props) {
    let i = 0;
    while (i < props.x) {
      console.log(i);
      i += 1;
    }
  }
  `;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize do-while statements', async () => {
    const code = `
  function Example(props) {
    let i = 0;
    do {
      console.log(i);
      i += 1;
    } while (i < props.x)
  }
  `;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize switch statements', async () => {
    const code = `
  function Example(props) {
    switch (props.type) {
      case 'a':
        return examples.a(props.value);
      case 'b':
        return examples.b(props.value);
      case 'c':
        return examples.c(props.value);
      default:
        return examples.default(props.value);
    }
  }
  `;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize if statements', async () => {
    const code = `
  function Example(props) {
    if (props.type === 'a') {
      return examples.a(props.value);
    } else {
      return examples.b(props.value);
    }
  }
  `;
    expect(await compile(code)).toMatchSnapshot();
  });
  it('should optimize try statements', async () => {
    const code = `
  function Example(props) {
    try {
      props.a();
    } catch (e) {
      props.b(e);
    } finally {
      props.c();
    }
  }
  `;
    expect(await compile(code)).toMatchSnapshot();
  });
});
