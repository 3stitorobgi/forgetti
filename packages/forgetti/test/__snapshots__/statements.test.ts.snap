// Vitest Snapshot v1

exports[`statements > should optimize do-while statements 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
function Example(props) {
  let _c2 = _useMemo(() => new Array(1), []);
  let i = 0;
  let _c = _c2[0] || (_c2[0] = []),
    _id = 0;
  do {
    let _lid = _id++,
      _l = _c[_lid] || (_c[_lid] = new Array(2));
    let _eq = Object.is(_l[0], i),
      _v = _eq ? _l[0] : _l[0] = i;
    let _v2 = _eq ? _l[1] : _l[1] = console.log(_v);
    _v2;
    let _v3 = i += 1;
    _v3;
  } while (i < props.x);
}"
`;

exports[`statements > should optimize for statements 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
function Example(props) {
  let _c2 = _useMemo(() => new Array(1), []);
  let _c = _c2[0] || (_c2[0] = []),
    _id = 0;
  for (let i = 0; i < 10; i += 1) {
    let _lid = _id++,
      _l = _c[_lid] || (_c[_lid] = new Array(2));
    let _eq = Object.is(_l[0], i),
      _v = _eq ? _l[0] : _l[0] = i;
    let _v2 = _eq ? _l[1] : _l[1] = console.log(_v);
    _v2;
  }
}"
`;

exports[`statements > should optimize for-in statements 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
function Example(props) {
  let _c = _useMemo(() => new Array(3), []);
  let _eq = Object.is(_c[0], props),
    _v = _eq ? _c[0] : _c[0] = props;
  let _v2 = _eq ? _c[1] : _c[1] = _v.arr;
  let _c2 = _c[2] || (_c[2] = []),
    _id = 0;
  for (const x in _v2) {
    let _lid = _id++,
      _l = _c2[_lid] || (_c2[_lid] = new Array(2));
    let _eq2 = Object.is(_l[0], x),
      _v3 = _eq2 ? _l[0] : _l[0] = x;
    let _v4 = _eq2 ? _l[1] : _l[1] = console.log(_v3);
    _v4;
  }
}"
`;

exports[`statements > should optimize for-of statements 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
function Example(props) {
  let _c = _useMemo(() => new Array(3), []);
  let _eq = Object.is(_c[0], props),
    _v = _eq ? _c[0] : _c[0] = props;
  let _v2 = _eq ? _c[1] : _c[1] = _v.arr;
  let _c2 = _c[2] || (_c[2] = []),
    _id = 0;
  for (const x of _v2) {
    let _lid = _id++,
      _l = _c2[_lid] || (_c2[_lid] = new Array(2));
    let _eq2 = Object.is(_l[0], x),
      _v3 = _eq2 ? _l[0] : _l[0] = x;
    let _v4 = _eq2 ? _l[1] : _l[1] = console.log(_v3);
    _v4;
  }
}"
`;

exports[`statements > should optimize if statements 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
function Example(props) {
  let _c = _useMemo(() => new Array(5), []);
  let _eq = Object.is(_c[0], props),
    _v = _eq ? _c[0] : _c[0] = props;
  let _v2 = _eq ? _c[1] : _c[1] = _v.type;
  let _eq2 = Object.is(_c[2], _v2),
    _v3 = _eq2 ? _c[2] : _c[2] = _v2;
  if (_v3 === 'a') {
    let _c2 = _c[3] || (_c[3] = new Array(3));
    let _v4 = _eq ? _c2[0] : _c2[0] = _v.value;
    let _eq3 = Object.is(_c2[1], _v4),
      _v5 = _eq3 ? _c2[1] : _c2[1] = _v4;
    let _v6 = _eq3 ? _c2[2] : _c2[2] = examples.a(_v5);
    return _v6;
  } else {
    let _c3 = _c[4] || (_c[4] = new Array(3));
    let _v7 = _eq ? _c3[0] : _c3[0] = _v.value;
    let _eq4 = Object.is(_c3[1], _v7),
      _v8 = _eq4 ? _c3[1] : _c3[1] = _v7;
    let _v9 = _eq4 ? _c3[2] : _c3[2] = examples.b(_v8);
    return _v9;
  }
}"
`;

exports[`statements > should optimize switch statements 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
function Example(props) {
  let _c = _useMemo(() => new Array(6), []);
  let _eq = Object.is(_c[0], props),
    _v = _eq ? _c[0] : _c[0] = props;
  let _v2 = _eq ? _c[1] : _c[1] = _v.type;
  switch (_v2) {
    case 'a':
      let _c2 = _c[2] || (_c[2] = new Array(3));
      let _v3 = _eq ? _c2[0] : _c2[0] = _v.value;
      let _eq2 = Object.is(_c2[1], _v3),
        _v4 = _eq2 ? _c2[1] : _c2[1] = _v3;
      let _v5 = _eq2 ? _c2[2] : _c2[2] = examples.a(_v4);
      return _v5;
    case 'b':
      let _c3 = _c[3] || (_c[3] = new Array(3));
      let _v6 = _eq ? _c3[0] : _c3[0] = _v.value;
      let _eq3 = Object.is(_c3[1], _v6),
        _v7 = _eq3 ? _c3[1] : _c3[1] = _v6;
      let _v8 = _eq3 ? _c3[2] : _c3[2] = examples.b(_v7);
      return _v8;
    case 'c':
      let _c4 = _c[4] || (_c[4] = new Array(3));
      let _v9 = _eq ? _c4[0] : _c4[0] = _v.value;
      let _eq4 = Object.is(_c4[1], _v9),
        _v10 = _eq4 ? _c4[1] : _c4[1] = _v9;
      let _v11 = _eq4 ? _c4[2] : _c4[2] = examples.c(_v10);
      return _v11;
    default:
      let _c5 = _c[5] || (_c[5] = new Array(3));
      let _v12 = _eq ? _c5[0] : _c5[0] = _v.value;
      let _eq5 = Object.is(_c5[1], _v12),
        _v13 = _eq5 ? _c5[1] : _c5[1] = _v12;
      let _v14 = _eq5 ? _c5[2] : _c5[2] = examples.default(_v13);
      return _v14;
  }
}"
`;

exports[`statements > should optimize try statements 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
function Example(props) {
  let _c2 = _useMemo(() => new Array(3), []);
  try {
    let _c = _c2[0] || (_c2[0] = new Array(2));
    let _eq = Object.is(_c[0], props),
      _v = _eq ? _c[0] : _c[0] = props;
    let _v2 = _eq ? _c[1] : _c[1] = _v.a();
    _v2;
  } catch (e) {
    let _c3 = _c2[1] || (_c2[1] = new Array(3));
    let _eq2 = Object.is(_c3[0], props),
      _v3 = _eq2 ? _c3[0] : _c3[0] = props;
    let _eq3 = Object.is(_c3[1], e),
      _v4 = _eq3 ? _c3[1] : _c3[1] = e;
    let _eq4 = _eq2 && _eq3,
      _v5 = _eq4 ? _c3[2] : _c3[2] = _v3.b(_v4);
    _v5;
  } finally {
    let _c4 = _c2[2] || (_c2[2] = new Array(2));
    let _eq5 = Object.is(_c4[0], props),
      _v6 = _eq5 ? _c4[0] : _c4[0] = props;
    let _v7 = _eq5 ? _c4[1] : _c4[1] = _v6.c();
    _v7;
  }
}"
`;

exports[`statements > should optimize while statements 1`] = `
"import { useMemo as _useMemo } from \\"react\\";
function Example(props) {
  let _c2 = _useMemo(() => new Array(1), []);
  let i = 0;
  let _c = _c2[0] || (_c2[0] = []),
    _id = 0;
  while (i < props.x) {
    let _lid = _id++,
      _l = _c[_lid] || (_c[_lid] = new Array(2));
    let _eq = Object.is(_l[0], i),
      _v = _eq ? _l[0] : _l[0] = i;
    let _v2 = _eq ? _l[1] : _l[1] = console.log(_v);
    _v2;
    let _v3 = i += 1;
    _v3;
  }
}"
`;
